### ISSUE TITLE
ShadowDOM with AMD broken?

### ISSUE THREAD

**Issue by DrNiels** [microsoft/monaco-editor#1843]
**monaco-editor version:** 0.20.0
**Browser:** Chrome
**OS:** Windows
**Codepen that shows the issue:** https://codepen.io/DrNiels/pen/xxGEjYX

I'm glad to see that Monaco can be used in a ShadowDom now. Unfortunately, I didn't get it to work. I followed steps very similar to the AMD sample (https://github.com/microsoft/monaco-editor-samples/tree/master/browser-amd-editor), but multiple aspects are broken:
- The blinking cursor does not move visually when clicking anywhere or writing text
- Scrolling breaks the displayed code
- Line numbers are not shown
- The column where the line numbers should be, ranges into the header

I'm not sure if the issues are due to the AMD loading, but as you have tests for ShadowDOMs that utilize webpack, I thought that could be the difference.

I hope I didn't do some dumb mistake, but the posted Codepen is mostly Copy + Paste from the sample plus some ShadowDOM adjustments.  When I remove the ShadowDOM part, everything works as intended

**Comment by alexdima** [comment]
@DrNiels There is one burden on you which the editor does not handle automatically because it depends from integrator to integrator and from AMD to ESM, etc.

The extra burden is that you need to make sure you load somehow the editor's CSS inside the shadow root. In your sample, the `editor.main.css` is loaded, but it is loaded in the global document. You would need some more code to copy those styles into the shadow root.

But not even that is sufficient, because font families cannot be defined in the shadow root, so you would need to define our codicon font in the global document.

Here is a sample:

https://github.com/microsoft/monaco-editor/blob/daabff0021fd6d4f5b57479b99b9a8ab900f951a/test/shadow-dom.html#L27-L43

Please let me know if you have any ideas about how this could be simplified for you as an integrator...

**Comment by cintaccs** [comment]
I have spend a lot of time trying to get the editor into a web component. Inspired by https://github.com/vanillawc/wc-monaco-editor which unfortunately is on 18.1 only - and did not initially fit what I needed - I tried to get the editor working in a lit-element https://lit-element.polymer-project.org/.

I get similar result as @DrNiels - I have tried the suggestions from the test/shadow-dom.html but haven't been able to get it right. The combination of workers - self environment - web-pack build and web-component shadow dom is simply too complex.

I can get it to work kind of ok - in the normal DOM with lit-element rendering to the non-shadow DOM root - however as soon as you combine multiple shadow DOM components - it gets hard to achieve that combination. I am almost at the stage where I am considering going back to an IFRAME implementation of monaco - which however has other disadvantages (more difficult to work with mutiple instances on same DOM page etc. - it is  possible - but gets hard to follow). Ideally I would love to see a web-component npm version of monaco-editor which could have all the attributes in the custom tag as well as it could have a promise fetch event that could be listened to by a parent component to deliver more complex model objects etc.

Having all the styles specific for monaco isolated in a web components shadow DOM as well as hiding all the underlying complexity of web workers etc. in the component seems ideal as far as I can see. IMO monaco editor as a self served web-component would be a very popular feature that could be used everywhere an editor is needed. In cases where multiple instances are needed - the web-component would allow that too as far as I can see.

It might be my lack of skills not being able to achieve this - but I have tried probably 1-2 weeks all together over the last years to achieve better encapsulation of this great editor in my application.

**Comment by DrNiels** [comment]
Thanks for the fast response! I fixed it by just loading the editor.main.css into my shadowDOM. In my productive system, this seems to completely fix the issue. My minimal example on CodePen still has the issue that the marker position is not displayed correctly: https://codepen.io/DrNiels/pen/poJREQo

To make this process simpler for developers, maybe you could add examples with ShadowDOM implementation into the monaco-editor-samples? Another helpful feature could be that you check if the required styles are available when monaco.editor.create is called and provide helpful error messages if something is missing?

**Comment by alexdima** [comment]
I have added a sample at https://github.com/microsoft/monaco-editor-samples/tree/master/browser-amd-shadow-dom

This uses the following tricks:
- imports codicon.ttf manually outside the shadow dom
- uses `'vs/css': { disabled: true }` to disable that the loader imports css globally
- imports css inside the shadow dom manually via `@import` directive
